### ISSUE TITLE
Upgrade failed when upgrading a TPR resource

### ISSUE THREAD

**Issue by lachie83** [kubernetes/helm#1531]
Using the chart in this PR https://github.com/kubernetes/charts/pull/214

Trying to simulate an etcd upgrade by deloying a TRP resource with etcd version v3.0.14 then upgrading using the default cluster.version from values.yaml to perform an upgrade.  Not sure if this is related to https://github.com/kubernetes/kubernetes/issues/29542

```
helm install . --name etcd-operator --set cluster.version=v3.0.14
helm upgrade etcd-operator .
Error: UPGRADE FAILED: unable to find api field in struct Unknown for the json field "metadata"
```

```
helm version
Client: &version.Version{SemVer:"v2.0.0-rc.2", GitCommit:"f8037c83c647b56d9ef42d0d5254e5fc563e3048", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.0.0-rc.2", GitCommit:"f8037c83c647b56d9ef42d0d5254e5fc563e3048", GitTreeState:"clean"}
```

**Comment by technosophos** [comment]
Out of curiosity, can you dump the upgrade YAML, and then run it manually with `kubectl apply`? We've been trying to trace down some issues with the apply logic.

**Comment by lachie83** [comment]
```
lachlanevenson@faux$ helm install . --dry-run --debug --name etcd-operator --set cluster.version=v3.0.14
Created tunnel using local port: '50430'
SERVER: ":50430"
NAME:   etcd-operator
TARGET NAMESPACE:   default
CHART:  etcd-operator 0.1.0
MANIFEST: 
---
# Source: etcd-operator/templates/deployment.yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: etcd-operator-etcd-opera
  labels:
    chart: "etcd-operator-0.1.0"
    app: etcd-operator-etcd-opera
    heritage: "Tiller"
    release: "etcd-operator"
spec:
  replicas: 1
  template:
    metadata:
      labels:
        chart: "etcd-operator-0.1.0"
        app: etcd-operator-etcd-opera
        heritage: "Tiller"
        release: "etcd-operator"
    spec:
      containers:
      - name: etcd-operator
        image: "quay.io/coreos/etcd-operator:latest"
        imagePullPolicy: IfNotPresent
        command: 
        - "/bin/sh"
        - "-c"
        - "/usr/local/bin/etcd-operator --pv-provisioner=kubernetes.io/gce-pd"
        env:
        - name: MY_POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        resources:
            limits:
              cpu: 100m
              memory: 128Mi
            requests:
              cpu: 100m
              memory: 128Mi
            
---
# Source: etcd-operator/templates/etcd-cluster.yaml
apiVersion: "coreos.com/v1"
kind: "EtcdCluster"
metadata:
  labels:
    chart: "etcd-operator-0.1.0"
    app: etcd-operator-etcd-opera
    heritage: "Tiller"
    release: "etcd-operator"
  name: "etcd-cluster"
spec:
  size: 3
  version: "v3.0.14"
  backup:
    maxSnapshot: 5
    snapshotIntervalInSecond: 30
    storageType: PersistentVolume
    volumeSizeInMB: 512
```

**Comment by lachie83** [comment]
Manual `kubectl apply` worked. FYI the manifest was rendered by hand

```
lachlanevenson@faux$ cat etcd-cluster-upgrade.yaml 
apiVersion: "coreos.com/v1"
kind: "EtcdCluster"
metadata:
  labels:
    chart: "etcd-operator-0.1.0"
    app: "etcd-operator-etcd-opera"
    heritage: "Tiller"
    release: "etcd-operator"
  name: "etcd-cluster"
spec:
  size: 3
  version: "v3.1.0-rc.0"
  backup:
    maxSnapshot: 5
    snapshotIntervalInSecond: 30
    storageType: PersistentVolume
    volumeSizeInMB: 512
lachlanevenson@faux$ kubectl apply -f etcd-cluster-upgrade.yaml 
etcdcluster "etcd-cluster" configured
```

**Comment by lachie83** [comment]
Here's the installed resource on Kubernetes

```
lachlanevenson@faux$ kubectl get Etcdcluster -o yaml
apiVersion: v1
items:
- apiVersion: coreos.com/v1
  data: eyJhcGlWZXJzaW9uIjoiY29yZW9zLmNvbS92MSIsImtpbmQiOiJFdGNkQ2x1c3RlciIsIm1ldGFkYXRhIjp7ImxhYmVscyI6eyJhcHAiOiJldGNkLW9wZXJhdG9yLWV0Y2Qtb3BlcmEiLCJjaGFydCI6ImV0Y2Qtb3BlcmF0b3ItMC4xLjAiLCJoZXJpdGFnZSI6IlRpbGxlciIsInJlbGVhc2UiOiJldGNkLW9wZXJhdG9yIn0sIm5hbWUiOiJldGNkLWNsdXN0ZXIifSwic3BlYyI6eyJiYWNrdXAiOnsibWF4U25hcHNob3QiOjUsInNuYXBzaG90SW50ZXJ2YWxJblNlY29uZCI6MzAsInN0b3JhZ2VUeXBlIjoiUGVyc2lzdGVudFZvbHVtZSIsInZvbHVtZVNpemVJbk1CIjo1MTJ9LCJzaXplIjozLCJ2ZXJzaW9uIjoidjMuMS4wLXJjLjAifX0=
  kind: EtcdCluster
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: '{"apiVersion":"extensions/v1beta1","metadata":{"name":"etcd-cluster","namespace":"default","creationTimestamp":null,"labels":{"app":"etcd-operator-etcd-opera","chart":"etcd-operator-0.1.0","heritage":"Tiller","release":"etcd-operator"}},"data":"eyJhcGlWZXJzaW9uIjoiY29yZW9zLmNvbS92MSIsImtpbmQiOiJFdGNkQ2x1c3RlciIsIm1ldGFkYXRhIjp7ImxhYmVscyI6eyJhcHAiOiJldGNkLW9wZXJhdG9yLWV0Y2Qtb3BlcmEiLCJjaGFydCI6ImV0Y2Qtb3BlcmF0b3ItMC4xLjAiLCJoZXJpdGFnZSI6IlRpbGxlciIsInJlbGVhc2UiOiJldGNkLW9wZXJhdG9yIn0sIm5hbWUiOiJldGNkLWNsdXN0ZXIifSwic3BlYyI6eyJiYWNrdXAiOnsibWF4U25hcHNob3QiOjUsInNuYXBzaG90SW50ZXJ2YWxJblNlY29uZCI6MzAsInN0b3JhZ2VUeXBlIjoiUGVyc2lzdGVudFZvbHVtZSIsInZvbHVtZVNpemVJbk1CIjo1MTJ9LCJzaXplIjozLCJ2ZXJzaW9uIjoidjMuMS4wLXJjLjAifX0="}'
    creationTimestamp: 2016-11-11T19:19:57Z
    labels:
      app: etcd-operator-etcd-opera
      chart: etcd-operator-0.1.0
      heritage: Tiller
      release: etcd-operator
    name: etcd-cluster
    namespace: default
    resourceVersion: "148095"
    selfLink: /apis/coreos.com/v1/namespaces/default/etcdclusters/etcd-cluster
    uid: d57def06-a843-11e6-a53f-42010a8a0ff3
  spec:
    backup:
      maxSnapshot: 5
      snapshotIntervalInSecond: 30
      storageType: PersistentVolume
      volumeSizeInMB: 512
    size: 3
    version: v3.0.14
kind: List
metadata: {}
```

**Comment by lachie83** [comment]
Seeing this post upgrade to release 2.0
`2016-11-17 17:39:24.812002 I | release_server.go:323: warning: Release Upgrade "etcd-operator" failed: extensions.ThirdPartyResourceData is not suitable for converting to "coreos.com/v1"
`

**Comment by technosophos** [comment]
For the time being, we may be stuck recommending that people recreate TPR resources or manually upgrade until upstream finishes support for this.

Let me know if you find a viable workaround.

**Comment by lachie83** [comment]
I'm happy with that. Thanks @technosophos . No work arounds as of yet.

**Comment by technosophos** [comment]
Has anyone tested this with Kubernetes 1.5?

**Comment by lachie83** [comment]
I will test with 1.5

**Comment by lachie83** [comment]
Different issue in 1.5 running Helm 2.1.0.

I can however apply the resulting manifests from the dry-run output manually

```
helm5 install . --name etcd-operator --set cluster.enabled=true --dry-run --debug
Created tunnel using local port: '62221'
SERVER: "localhost:62221"
CHART PATH: /Users/lachlanevenson/sandbox/github/kubernetes/charts/stable/etcd-operator
NAME:   etcd-operator
TARGET NAMESPACE:   default
CHART:  etcd-operator 0.1.0
MANIFEST: 
---
# Source: etcd-operator/templates/deployment.yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: etcd-operator-etcd-opera
  labels:
    chart: "etcd-operator-0.1.0"
    app: "etcd-operator-etcd-opera"
    heritage: "Tiller"
    release: "etcd-operator"
spec:
  replicas: 1
  template:
    metadata:
      labels:
        chart: "etcd-operator-0.1.0"
        app: etcd-operator-etcd-opera
        heritage: "Tiller"
        release: "etcd-operator"
    spec:
      containers:
      - name: etcd-operator
        image: "quay.io/coreos/etcd-operator:latest"
        imagePullPolicy: IfNotPresent
        command: 
        - "/bin/sh"
        - "-c"
        - "/usr/local/bin/etcd-operator --pv-provisioner=kubernetes.io/gce-pd"
        env:
        - name: MY_POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        resources:
            limits:
              cpu: 100m
              memory: 128Mi
            requests:
              cpu: 100m
              memory: 128Mi
            
---
# Source: etcd-operator/templates/etcd-cluster.yaml
apiVersion: "coreos.com/v1"
kind: "EtcdCluster"
metadata:
  labels:
    chart: "etcd-operator-0.1.0"
    app: "etcd-operator-etcd-opera"
    heritage: "Tiller"
    release: "etcd-operator"
  name: "etcd-cluster"
spec:
  size: 3
  version: "v3.1.0-rc.0"
lachlanevenson@faux$ helm5 install . --name etcd-operator --set cluster.enabled=true
Error: release etcd-operator failed: unable to decode "": no kind "EtcdCluster" is registered for version "coreos.com/v1"
```

**Comment by technosophos** [comment]
See related ticket #1468

**Comment by technosophos** [comment]
We think that the problem above is that you cannot create the TPR and an instance both at the same time. Maybe you could try using a pre-install hook to install the Deployment?

Also, we seem to be hitting some differences between k8s 1.4.7 and 1.5.1 (specifically with `kubectl apply` on TPRs). Still testing that.

**Comment by enisoc** [comment]
> you cannot create the TPR and an instance both at the same time

Do you mean that Kubernetes doesn't like this, or just that there's an interaction with the way Helm works? If the former, why does it work when applying the dry-run manifest directly with kubectl?

**Comment by technosophos** [comment]
@enisoc The Etcd one is a good example: If you supply both the Deployment and the EtcdCluster resources in the same chart, both will be supplied to the Kubernetes API for validation at the same time. Since the Deployment installs the TPR controller, the "coreos.com/v1" group won't be defined yet. So the entire release will fail.

The dry-run succeeds because that data is never submitted to the Kube API server.

The work-around is to just install the TPR definition of "coreos.com/v1" in a pre-install hook so that it is already online when the main chart submits its EtcdCluster resources.

In the example above, I suggested that @lachie83 install the deployment as a pre-install hook since the controller needs to be up and running before a resource type of EtcdCluster will trigger a deployment. In other cases, it'd be fine to just pre-install the ThirdPartyResourceType manifest, and spin up the controller in the main chart.

**Comment by enisoc** [comment]
I'm still not sure that Kubernetes itself imposes the requirement that the installation of the TPR definition must precede creation of resources of that Kind.

When I said it works when applying the dry-run manifest, I meant: Copy and paste the output of `helm install --debug --dry-run ...` into a single YAML file and run `kubectl create -f` on it. Everything came up normally, so it seems that the problem only occurs when Helm is in the loop, unless that success was purely chance and the failure is racy. If that's what you're suggesting, I'll run some more tests to confirm.

**Comment by lachie83** [comment]
@enisoc The etcd-operator is creating the etcd-cluster.coreos.com TPR which is specified under the EtcdCluster resource apiVersion. If that hasn't been created you cannot submit the EtcdCluster resource kind. I would be interested to know whether the etcd-operator actually runs  in a failed helm deployment.

**Comment by lachie83** [comment]
not sure If I used the "workaround" (curl) to submit an update updated TPR but here's the Kubernetes issue https://github.com/kubernetes/kubernetes/issues/29542

**Comment by mgoodness** [comment]
I'm charting [kube-prometheus](https://github.com/coreos/kube-prometheus/issues/29), which (ideally) installs the Prometheus Operator and a slew of other manifests, then instantiates two TPRs that the Operator created. Running into the same issue as in #1739.

I tried @technosophos' suggestion of installing the Operator as a pre-install hook, to no avail. Message is `Error: apiVersion "monitoring.coreos.com/v1alpha1" in kube-prometheus/templates/alertmanager.yaml is not available`, which is the same as when I don't put the hook annotation on the Operator Deployment.

Split the Operator into its own chart and installed, waiting for it to create its TPRs. Then installed the other chart, only to get `Error: unable to decode "": no kind "Alertmanager" is registered for version "monitoring.coreos.com/v1alpha1"`. But like @lachie83, I am able to apply `--dry-run --debug` output.

**Comment by technosophos** [comment]
Okay, Helm 2.2.0 fixed the errors of the type `Error: unable to decode "": no kind "Alertmanager" is registered for version "monitoring.coreos.com/v1alpha1"`

However, we have no way of ensuring that just because a TPR installed into Kubernetes correctly, it is available for usage. So we don't have any other suggestions for installing TPRs other than to use `pre-install` hooks and hoping that Kubernetes bootstraps them fast enough. (Using the `--wait` flag on `helm install` might actually do the trick, though, since it slows down the install process)

Using a separate chart for defining the TPR seems to help, but only because it usually takes a human a long enough period of time to do two installs.

So at this point:

- we believe the TPR bugs are all fixed
- the length of time it takes Kubernetes to register a TPR is beyond our ability to influence
- your best bet is to use `pre-install` hooks plus `helm install --wait`

So I'm going to mark this as closed.

**Comment by errordeveloper** [comment]
I've encountered this kind of error outside of Helm, when I was experimenting with using `kubectl apply` on Istio CDRs. As this is the only issue where this error bubbles up when you google it, I thought I'd leave a comment on how to solve it.

I've installed Istio and the Bookinfo example. I've also created some configuration using `istioctl`, which is stored as CDRs in Kuberentes. I've then exported the config files with `kubectl get istioconfigs --export --output yaml` and using those as input to `kubectl apply` for the first time worked, but then I got this error:

```
extensions.ThirdPartyResourceData is not suitable for converting to "istio.io/v1alpha1"
```

When I examined my CDR files, I found that certain metadata fields (`creationTimestamp`, `resourceVersion`, `selfLink` and `uuid` were set). I've removed those metadata fields and `kubectl apply` after that.

Hope readers will find this comment useful!
